cinema/views.py

```diff
@@ -43,16 +45,100 @@ def get_serializer_class(self):
 
         return MovieSerializer
 
+    @staticmethod
+    def _params_to_int(query_string):
+        return [int(str_id) for str_id in query_string.split(",")]
+
+    def get_queryset(self):
+        queryset = self.queryset
+        actors = self.request.query_params.get("actors")
+        genres = self.request.query_params.get("genres")
+        title = self.request.query_params.get("title")
+
+        if actors:
+            actors = self._params_to_int(actors)
+            queryset = queryset.filter(actors__id__in=actors)
+
+        elif genres:
+            genres = self._params_to_int(genres)
+            queryset = queryset.filter(genres__id__in=genres)
+        elif title:
+            queryset = queryset.filter(title__icontains=title)
+        return queryset
+
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
     serializer_class = MovieSessionSerializer
 
+    @staticmethod
+    def _params_to_int(query_string):
+        return [int(str_id) for str_id in query_string.split(",")]
+
     def get_serializer_class(self):
         if self.action == "list":
             return MovieSessionListSerializer
-
-        if self.action == "retrieve":
+        elif self.action == "retrieve":
             return MovieSessionDetailSerializer
 
         return MovieSessionSerializer
+
+    def get_queryset(self):
+        queryset = self.queryset
+        if self.action in "list":
+            queryset = queryset.select_related().annotate(
+                tickets_available=(
+                    F("cinema_hall__rows")
+                    * F("cinema_hall__seats_in_row")
+                    - Count("tickets")
+                )
+            )
+
+            date = self.request.query_params.get("date")
+            movie = self.request.query_params.get("movie")
+            if date and movie:
+                movie = self._params_to_int(movie)
+                queryset = queryset.filter(
+                    show_time__date=date,
+                    movie__id__in=movie
+                )
+            elif date:
+                queryset = queryset.filter(
+                    show_time__date=date
+                )
```

Comment:
Code Efficiency issue: It seems you are splitting the date to filter the queryset (lines 101 and 107). According to the checklist, this is not an efficient way to filter by date. Try filtering the date directly without splitting it. Refer to the checklist for a good example.